File: frontend/src/types/tool.ts
export interface Prerequisite {
  id: string;
  name: string;
  description: string;
  isInstalled: boolean;
  required: boolean;
  installNote: string;
}

export interface PrereqDoneEvent {
  id: string;
  success: boolean;
  error?: string;
}

export type Category =
  | "core"
  | "languages"
  | "databases"
  | "cloud"
  | "editors"
  | "terminal"
  | "dotfiles"
  | "apps";

export type PermissionLevel = "user" | "admin";

export type InstallStatus = "installed" | "failed" | "skipped";

export interface Tool {
  id: string;
  name: string;
  description: string;
  category: Category;
  section: string;
  permissionLevel: PermissionLevel;
  defaultEnabled: boolean;
  requiresSudo: boolean;
  isInstalled: boolean;
  version?: string;
}

export interface Profile {
  id: string;
  name: string;
  icon: string;
  toolIds: string[];
}

export interface Installation {
  id: number;
  toolId: string;
  toolName: string;
  status: InstallStatus;
  output: string;
  durationMs: number;
  installedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface InstallStartEvent {
  total: number;
  toolIds: string[];
}

export interface InstallProgressEvent {
  current: number;
  total: number;
  toolId: string;
  toolName: string;
}

export interface InstallLineEvent {
  toolId: string;
  line: string;
  isStderr: boolean;
}

export interface InstallToolDoneEvent {
  toolId: string;
  success: boolean;
  durationMs: number;
}

export interface InstallDoneEvent {
  installed: number;
  failed: number;
  skipped: number;
}

File: internal/models/tool.go
package models

type Category string

const (
	CategoryCore      Category = "core"
	CategoryLanguages Category = "languages"
	CategoryDatabases Category = "databases"
	CategoryCloud     Category = "cloud"
	CategoryEditors   Category = "editors"
	CategoryTerminal  Category = "terminal"
	CategoryDotfiles  Category = "dotfiles"
	CategoryApps      Category = "apps"
)

type PermissionLevel string

const (
	PermissionUser  PermissionLevel = "user"
	PermissionAdmin PermissionLevel = "admin"
)

// Tool is a parsed registry entry. IsInstalled and Version are populated at runtime.
type Tool struct {
	ID              string          `json:"id"`
	Name            string          `json:"name"`
	Description     string          `json:"description"`
	Category        Category        `json:"category"`
	Section         string          `json:"section"`
	PermissionLevel PermissionLevel `json:"permissionLevel"`
	DefaultEnabled  bool            `json:"defaultEnabled"`
	RequiresSudo    bool            `json:"requiresSudo"`
	IsInstalled     bool            `json:"isInstalled"`
	Version         string          `json:"version,omitempty"`
}

// Profile is a preset bundle of tools.
type Profile struct {
	ID      string   `json:"id"`
	Name    string   `json:"name"`
	Icon    string   `json:"icon"`
	ToolIDs []string `json:"toolIds"`
}

type InstallStatus string

const (
	StatusInstalled InstallStatus = "installed"
	StatusFailed    InstallStatus = "failed"
	StatusSkipped   InstallStatus = "skipped"
)
